### ISSUE TITLE
Respect 'overrides' defined in dependencies

### ISSUE THREAD

**Issue by xaka** [taptapship/wiredep#149]
Let's say we have **Package A** with the following `bower.json`:

```
 "overrides": {
   "crypto-js": {
     "main": [
       "rollups/hmac-sha1.js",
       "components/enc-base64.js"
     ]
   }
 }
```

And **Package B** which uses **Package A** as a dependency. When I use `wiredep` inside **Package B** directory, I'm NOT getting files mentioned in `overrides` above.

At the same time when I use `wiredep` inside **Package A** directory, everything works as expected.

I'd say it's a bug because I have to maintain `overrides` section across multiple packages, which is very annoying for lazy people :)

```
package-a/
  bower.json
package-b/
  bower.json
```

**Comment by CWSpear** [comment]
Why are you running `wiredep` inside package directories? When you say "package directories," what precisely do you mean? Bower packages?

**Comment by xaka** [comment]
I mean the directory, which contains `bower.js` file.

**Comment by CWSpear** [comment]
But you are talking about multiple. I have never been on a project with multiple `bower.js` files, so I'm trying to understand.

**Comment by xaka** [comment]
No no, there is only one `bower.json`. Let's say you have project A which depends on library B which overrides `main` sections of its own dependencies. When you run `wiredep` for project A, it gives you back JS files of library B and its dependencies as expected, but at the same time it doesn't respect (and include) overrides defined by it.

```
project-a/
  bower.json (depedencies=library-b)

library-b/
  bower.json (dependencies=..., overrides=...)
```

But when I run `wiredep` for library B just for testing reasons, I'm getting back JS files of its dependencies and overrides.

**Comment by CWSpear** [comment]
Ok, I *think* I understand... so `overrides` is *not* a standard `bower` property (there's been a lot of discussion about it in this repo, as well as others that use it and a lot of discussion on the bower repo as well).

`wiredep` doesn't look at any other `bower.json`'s except for the one in the current directory (i.e. you're app's `bower.json`). There has been talk about packages having `overrides` on them, but it would require a *much* wider adoption of `overrides`, which we have not been able to happen.

Basically, a package (i.e. a package in `bower`) declaring its `overrides` property doesn't currently really do anything.

I'll see if I can find any of those threads about previous discussions later, but hopefully that at least gives you an overview of the state of things.

**Comment by xaka** [comment]
Thanks for explanation, it's all clear. The reason I raised the issue is that it worked for one case and I was expecting it to work for almost similar case, but it didn't, which means I have to duplicate overrides in my project for every dependency with overrides it's referencing to.

About using not standard properties. I don't think it's an issue as there are many of those for package.json for instance. It's the matter of keeping names unique so for wiredep instead of using `overrides` we could introduce property called `wiredep`. The overall idea behind `overrides` is great and it requires a little bit more effort (overrides inheritance) to move it to the next level.

**Comment by CWSpear** [comment]
The reason is *is* **overrides** is the implementation we used was compatible with other popular plugins out there. There are others, but I've personally used `wiredep` and [`gulp-bower-files`](https://www.npmjs.com/package/gulp-bower-files) together and it was nice just having to put `overrides` in one place instead of having to copy it multiple places.

But again, that all works cuz it's just your own app code and not a package out in the wild, which is where it can get tricky.

If you're bored and needed some light reading: bower/bower#585 and bower/bower.json-spec#27

**Comment by stephenplusplus** [comment]
This could be dangerous if your project has some dependency which overrides one of *its* dependency's `main` files, but doesn't consider that your project may have the same dependency. `wiredep` could try to handle this for you, but I don't think supporting this overall use case is a good idea for the reasons @CWSpear highlighted. Thanks for taking the time to help improve this project and walk us through the situation!
